src/nanotron/fp8/tensor.py

```diff
@@ -0,0 +1,106 @@
+import torch
+import transformer_engine as te  # noqa
+import transformer_engine_extensions as tex
+
+from nanotron.fp8.constants import DTYPE_TO_FP8_MAX, FP8_DTYPES, INITIAL_SCALING_FACTOR
+from nanotron.fp8.dtypes import DTypes
+
+
+class FP8Tensor(torch.Tensor):
+    """FP8 Tensor."""
+
+    def __new__(cls, tensor: torch.Tensor, dtype: DTypes) -> torch.Tensor:
+        # TODO(xrsrke): there is a circular import issue
+        # between tensor.py and meta.py fix this
+        from nanotron.fp8.meta import FP8Meta
+
+        # TODO(xrsrke): if the tensor is on cpu, then bypass the quantization
+        # because the current kernels only support gpu tensor
+        assert tensor.device != torch.device("cpu"), "FP8Tensor only supports CUDA device"
+        assert isinstance(dtype, DTypes)
+
+        amax = tensor.abs().max().clone()
+        scale = update_scaling_factor(amax, torch.tensor(INITIAL_SCALING_FACTOR, dtype=torch.float32), dtype)
+        fp8_meta = FP8Meta(amax, scale, dtype)
+
+        if tensor.dtype not in FP8_DTYPES:
+            fp8_tensor = convert_tensor_to_fp8(tensor, fp8_meta)
+        else:
+            fp8_tensor = tensor
```

Comment:
what does it mean to have `tensor` alrdy in fp8. I think that may create issues because we're overriding its fp8_meta. Let's assert `tensor` is not fp8

Reply:
good catch. thanks